app-backend/datamodel/src/main/scala/Scope.scala

```diff
@@ -0,0 +1,308 @@
+package com.rasterfoundry.datamodel
+
+import cats.{Eq, Monoid}
+import cats.implicits._
+import io.circe.{Decoder, DecodingFailure, Encoder, HCursor, Json}
+import io.circe.parser._
+
+import scala.util.{Failure, Success}
+
+case class Action(domain: String, action: String, limit: Option[Long]) {
```

Comment:
I'm not sure if this is related to your comment [below](https://github.com/raster-foundry/raster-foundry/pull/5270/files#r361752725) about ad hoc scopes, but I feel like the constructor should be a little more defensive here instead of accepting _any_ string. The decoder catches bad actions below, but it seems like more could be done here to help prevent them from being created to start with - especially ones with strings like `;` which could cause problems decoding other actions. I think if the only way to add/remove scopes is via `json` in the API we could be fine because the decoder will catch it.

One option would be `refined`. Is there a reason `action` and or `domain` could not be an `enum`?

I'm not sure if this is being too defensive though.

Reply:
The problem I couldn't think around is that while domain and action could be enums, it isn't the case that every action is valid in every domain. At the point where it's possible to create an `Action` with the `datasources` domain and a `createAnnotation` action, I don't think we get too much extra validity, and the type-level programming to prevent that is I think beyond my current understanding.

I'm also not sure that there will be API endpoints for adding scopes. Scopes should only change if:

- a regular user gets promoted to/demoted from a team or org admin
- a Groundwork user bumps from free to paid

We've deferred the discussion on what that transition should look like for a bit, so maybe we're getting close to time to figure that out.

I don't think this is related to my comment about ad hoc scopes -- what I meant by that is that you can remove any canned policy from the pattern match, and as long as the matches end with `ComplexScope(_)` and `SimpleScope(_)`, the compiler still (correctly) thinks the match is exhaustive. That would still be true with enums instead/refined types in the place of strings.